tests/components/mqtt/test_init.py

```diff
@@ -1342,6 +1342,84 @@ def _callback_b(msg: ReceiveMessage) -> None:
     assert len(calls_b) == 1
 
 
+async def test_replaying_payload_wildcard_topic(
+    hass: HomeAssistant,
+    mqtt_client_mock: MqttMockPahoClient,
+    mqtt_mock_entry_no_yaml_config: MqttMockHAClientGenerator,
+) -> None:
+    """Test replaying retained messages.
+
+    When we have multiple subscriptions to the same wildcard topic,
+    SUBSCRIBE must be sent to the broker again
+    for it to resend any retained messages for new subscriptions.
+    Retained messages should only be replayed for new subscriptions, except
+    when the MQTT client is reconnection.
+    """
+    mqtt_mock = await mqtt_mock_entry_no_yaml_config()
+
+    # Fake that the client is connected
+    mqtt_mock().connected = True
+
+    calls_a: list[ReceiveMessage] = []
+    calls_b: list[ReceiveMessage] = []
+
+    def _callback_a(msg: ReceiveMessage) -> None:
+        calls_a.append(msg)
+
+    def _callback_b(msg: ReceiveMessage) -> None:
+        calls_b.append(msg)
+
+    await mqtt.async_subscribe(hass, "test/#", _callback_a)
+    # Simulate (retained) messages being played back on new subscriptions
+    async_fire_mqtt_message(hass, "test/state1", "new_value_1", qos=0, retain=True)
+    async_fire_mqtt_message(hass, "test/state2", "new_value_2", qos=0, retain=True)
+    await hass.async_block_till_done()
+    assert len(calls_a) == 2
+    mqtt_client_mock.subscribe.assert_called()
+    calls_a = []
+    mqtt_client_mock.reset_mock()
+
+    # resubscribe to the wild card topic again
```

Comment:
I mean, the ideal would be to NOT re-process a second RETAIN=1 message to subscribers that already got it for a given topic.

eg.
Sub 1 of test/# should process ONE RETAIN=1 for test/state1 and ONE for test/state2

But Sub 2 of test/# (ie. a re-subscribe with a different wildcard subscriber) should not trigger Sub 1 to re-process a RETAIN=1 for test/state1 or test/state2 (though Sub 2 should process one for each).

However, the management of that is horrendous.
You essentially have to make a _retained_init variable like you have for non-wildcards, for each wildcard (so each wildcard subscription would track which fully resolved topics it received RETAIN=1 messages for already)
Meaning you would end up checking a self._wildcard_retained_init[Sub n][topic] rather than the self._retained_init[Sub n] you use for non-wildcards.

It's such a niche use case it's probably not worth the effort.  But TECHNICALLY, we are doing it wrong if we allow Sub 1 to process multiple RETAIN=1 messages for a given topic.

The main reason I highlight this is because this test case implies it's the correct behavior (it's not - ideally if implemented correctly, len(calls_a) on line 1389 should be == 0).

Reply:
I excluded wildcard topics, this test shows they still work and that the replaying works. Not more than that.

Reply:
It is correct you would need to cache all topics, not to forget that in most common cases I do not expect wildcard topics to be resubscribed by other clients. It would have the side effect of replaying all last topic values. In real live the integrations claim some wildcard space for their discovery. So let us not put in to much effort is my thought.

Reply:
I agree, it's probably not worth the effort.  But might be worth putting a comment above the implementation to the effect of what the correct behavior is according to the spec, and that this doesn't follow that, but it is deemed reasonable due to the infrequency of wildcard subs.

ie. document that we know this is not 100% compliant, but implemented it this way on purpose.